lib/Sema/CSSimplify.cpp

```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+/// a type.
+/// This function may be slow for deep class hierarchies and multiple protocol
+/// conformances, but it is invoked only after other constraint simplification
+/// rules fail.
+static DynamicCallableMethods
+getDynamicCallableMethods(Type type, ConstraintSystem &CS,
+                          const ConstraintLocatorBuilder &locator,
+                          bool &error) {
+  auto canType = type->getCanonicalType();
+  auto it = CS.DynamicCallableCache.find(canType);
+  if (it != CS.DynamicCallableCache.end()) return it->second;
+
+  // Calculate @dynamicCallable methods for composite types with multiple
+  // components (protocol composition types and archetypes).
+  auto calculateForComponentTypes =
+      [&](ArrayRef<Type> componentTypes) -> DynamicCallableMethods {
+    DynamicCallableMethods methods;
+    for (auto componentType : componentTypes) {
+      auto tmp = getDynamicCallableMethods(componentType, CS, locator, error);
+      if (error) return methods;
+      if (tmp.argumentsMethod) {
+        if (methods.argumentsMethod &&
+            methods.argumentsMethod != tmp.argumentsMethod) {
+          error = true;
+          return methods;
+        }
+        methods.argumentsMethod = tmp.argumentsMethod;
+      }
+      if (tmp.keywordArgumentsMethod) {
+        if (methods.keywordArgumentsMethod &&
+            methods.keywordArgumentsMethod != tmp.keywordArgumentsMethod) {
+          error = true;
+          return methods;
+        }
+        methods.keywordArgumentsMethod = tmp.keywordArgumentsMethod;
+      }
+    }
+    return methods;
+  };
+
+  // Calculate @dynamicCallable methods.
+  auto calculate = [&]() -> DynamicCallableMethods {
+    // If this is an archetype type, check if any types it conforms to
+    // (superclass or protocols) have the attribute.
+    if (auto archetype = dyn_cast<ArchetypeType>(canType)) {
+      SmallVector<Type, 2> componentTypes;
+      for (auto protocolDecl : archetype->getConformsTo())
+        componentTypes.push_back(protocolDecl->getDeclaredType());
+      if (auto superclass = archetype->getSuperclass())
+        componentTypes.push_back(superclass);
+      return calculateForComponentTypes(componentTypes);
+    }
+
+    // If this is a protocol composition, check if any of its members have the
+    // attribute.
+    if (auto protocolComp = dyn_cast<ProtocolCompositionType>(canType))
+      return calculateForComponentTypes(protocolComp->getMembers());
+
+    // Otherwise, this must be a nominal type.
+    // Dynamic calling doesn't work for tuples, etc.
+    auto nominal = canType->getAnyNominal();
+    if (!nominal) return DynamicCallableMethods();
+
+    // If this type conforms to a protocol which has the attribute, then
+    // look up the methods.
+    for (auto p : nominal->getAllProtocols())
+      if (p->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
```

Comment:
Currently, `for (auto p : nominal->getAllProtocols())` iterates over all protocols and looks up methods in the first protocol that has the `@dynamicCallable` attribute.

Instead, should `calculateForComponentTypes` be called on the all protocols with the `@dynamicCallable` attribute? That seems more correct.